examples/palace-tonegen.py

```diff
@@ -0,0 +1,127 @@
+# or (at your option) any later version.
+#
+# palace is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with palace.  If not, see <https://www.gnu.org/licenses/>.
+
+from argparse import Action, ArgumentParser
+from array import array
+from math import pi, sin
+from random import random
+from typing import Iterable, Tuple
+
+from palace import Buffer, Context, BaseDecoder, Device
+
+from numpy import linspace
+from numpy.random import random_sample
+from scipy import signal
+
+PERIOD: float = 0.025
+WAVEFORMS = {'sine': sine,
+             'square': square,
+             'sawtooth': sawtooth,
+             'triangle': triangle,
+             'impulse': impulse,
+             'white noise': white_noise}
+
+
+class ToneGenerator(BaseDecoder):
+    def __init__(self, waveform: str):
+        self.func = lambda frame: WAVEFORMS[waveform](frame/self.frequency,
+                                           self.frequency)
+
+    @BaseDecoder.frequency.getter
+    def frequency(self) -> int: return 44100
+
+    @BaseDecoder.channel_config.getter
+    def channel_config(self) -> str:
+        return 'Mono'
+
+    @BaseDecoder.sample_type.getter
+    def sample_type(self) -> str:
+        return 'Float32'
+
+    @BaseDecoder.length.getter
+    def length(self) -> int: return 0
+
+    def seek(self, pos: int) -> bool: return False
+
+    @BaseDecoder.loop_points.getter
+    def loop_points(self) -> Tuple[int, int]: return 0, 0
+
+    def read(self, count: int) -> bytes:
+        stop = min(self.start + count, self.length)
+        data = array('f', map(self.func, range(self.start, stop)))
+        self.start = stop
+        return data.tobytes()
+
+
+class TypePrinter(Action):
+    def __call__(self, parser: ArgumentParser, *args, **kwargs) -> None:
+        print('Available waveform types:', *WAVEFORMS, sep='\n')
+        parser.exit()
+
+
+def sine(time: int, sample_rate: int) -> Iterable[float]:
+    """Generate sinusoidal signal."""
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = sin(t)
+    return data
+
+
+def square(time: int, sample_rate: int) -> Iterable[float]:
+    """Generate square signal."""
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = signal.square(2 * pi * 5 * t)
+
+
+def sawtooth(time: int, sample_rate: int) -> Iterable[float]:
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = signal.sawtooth(2 * pi * 5 * t)
+
+
+def triangle(time: int, sample_rate: int) -> Iterable[float]:
+    pass
+
+
+def impulse(time: int, sample_rate: int) -> Iterable[float]:
+    pass
+
+
+def white_noise(time: int, sample_rate: int) -> Iterable[float]:
+    return random_sample((1, time * sample_rate))
```

Comment:
These should be `(float) -> float` instead, i.e. for most you don't even have to write anything.

Reply:
Hmm, isn't time and sample rate two different parameters? Btw, is it ok not to put the type hinting for WAVEFORMS, since when I put it flake8 says EOF in multiline or smt.

Reply:
Somehow I missed this ugh.

> Hmm, isn't time and sample rate two different parameters?

Well yes but actually no, since you'll divide them for each other anyway.

> since when I put it flake8 says EOF in multiline or smt

Could be a typo.

Reply:
> Could be a typo.

True. I missed a `]` in all those clumsy brackets.